### PR TITLE
SageMaker Pipe Mode reader and build integration with setuptools

### PR DESCRIPTION
**Pull request by owen-t**
This is merging on my fork only, changes will not flow into main TF container.

*Description of changes:*
See commit.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cpp/pipemode/PipeReader/PipeReader.hpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            return channelDirectory;
+        }
+        std::string GetChannelName() const {
+            return channelName;
+        }
+
+    private:
+        std::string BuildCurrentPipeName() const;
+        void Open();
+        unsigned int currentPipeIndex;
```
Comment (nadiaya) [inline cpp/pipemode/PipeReader/PipeReader.hpp]: nit: can we use fixed width integer types: http://en.cppreference.com/w/cpp/types/integer ?

File: cpp/pipemode/PipeReader/PipeReader.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+void PipeReader::Close() {
+    if(-1 != currentPipe) {
+        close(currentPipe);
+        currentPipe = -1;
+    }
+}
+
+void PipeReader::Reset() {
+    Close();
+    currentPipeIndex ++;
```
Comment (nadiaya) [inline cpp/pipemode/PipeReader/PipeReader.cpp]: nit: pre-increment would be preferable in this case.

File: cpp/pipemode/PipeReader/PipeReader.hpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        */
+        void Close();
+        
+        /**
+           Returns the directory where channel pipes are read.
+        */
+        std::string GetChannelDirectory() const {
+            return channelDirectory;
+        }
+        std::string GetChannelName() const {
```
Comment (nadiaya) [inline cpp/pipemode/PipeReader/PipeReader.hpp]: Add comment for GetChannelName

**Review by nadiaya — APPROVED** [review]
LGTM

File: cpp/pipemode/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (mvsusp) [inline cpp/pipemode/CMakeLists.txt:1]: why not using 3.10 for both makelists?

File: cpp/pipemode/PipeReader/PipeReader.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            currentPipeIndex(0),
+            currentPipe(-1) {
+    Open();
+}
+
+PipeReader::~PipeReader() {
+    Close();
+}
+
+void PipeReader::Open() {
```
Comment (mvsusp) [inline cpp/pipemode/PipeReader/PipeReader.cpp:40]: Should we have a mutex for when we open a new stream?

File: cpp/pipemode/PipeReader/PipeReader.cpp (hunk 2 above)
Comment (mvsusp) [reply cpp/pipemode/PipeReader/PipeReader.cpp]: @nadiaya, would you elaborate the reason of that? For learning purposes.

File: cpp/pipemode/test/testPipeReader/TestPipeReader.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+#include <cstdlib>
+#include <iostream>
+#include <fstream>
+
+std::string createTemporaryDirectory() {
+    char mkdTemplate[] = "/tmp/tmpdir.XXXXXX";
+    return std::string(mkdtemp(mkdTemplate));
+}
+
+void createChannel(const std::string& channelDirectory, const std::string& channelName, const char* data, unsigned int index) {
```
Comment (mvsusp) [inline cpp/pipemode/test/testPipeReader/TestPipeReader.cpp]: Method is too long.

File: cpp/pipemode/test/testPipeReader/TestPipeReader.cpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    // Close the current pipe, without reading anything from it
+    reader.Close();
+
+    // Create second iteration of elizabeth
+    createChannel(channelDirectory, "elizabeth", "def", 1);
+    reader.Reset();
+    size_t readNum = reader.Read(static_cast<void*>(buffer), 4);
+    EXPECT_STREQ("def", buffer);
+    EXPECT_EQ(4, readNum);
+}
```
Comment (mvsusp) [inline cpp/pipemode/test/testPipeReader/TestPipeReader.cpp:139]: Can we have tests for the case where two readers are trying to read from the same stream?

**Review by mvsusp — COMMENTED** [review]
The design looks good to me.
We need tests for exceptions.
Can we have documentation on how to build and run the tests?

File: cpp/pipemode/PipeReader/PipeReader.cpp (hunk 2 above)
Comment (owen-t) [reply cpp/pipemode/PipeReader/PipeReader.cpp]: https://stackoverflow.com/questions/17366847/what-is-the-difference-between-pre-increment-and-post-increment-in-the-cycle-fo

File: cpp/pipemode/PipeReader/PipeReader.cpp (hunk 5 above)
Comment (owen-t) [reply cpp/pipemode/PipeReader/PipeReader.cpp:40]: Concurrency control will be in the TensorFlow iterator.

File: cpp/pipemode/test/testPipeReader/TestPipeReader.cpp (hunk 7 above)
Comment (owen-t) [reply cpp/pipemode/test/testPipeReader/TestPipeReader.cpp:139]: No, I'll add a comment saying that behavior is not supported. It won't happen in practice.

File: cpp/pipemode/CMakeLists.txt (hunk 4 above)
Comment (owen-t) [reply cpp/pipemode/CMakeLists.txt:1]: should be. I'll fix this in a subsequent PR.

**Pull request merged**